[Concurrency] Emit warning suggesting async function

etcwilde opened the pull request:
This patch adds a diagnostic to recommend using the async function
instead of using the completion handler version. The diagnostic is only
emitted when the call is made in an async context and the
completion-handler version of the function is attributed with the
@completionHandleyAsync attribute.

Resolves: rdar://75221638

etcwilde committed:
Emit warning suggesting async function

This patch adds a diagnostic to recommend using the async function
instead of using the completion handler version. The diagnostic is only
emitted when the call is made in an async context and the
completion-handler version of the function is attributed with the
@completionHandleyAsync attribute.

etcwilde commented:
@swift-ci please smoke test

ktoso reviewed (approved):
Nice one!

Would also offering a a fixitReplace be a bit too hard?

etcwilde commented:
> Nice one!
Thanks!

> Would also offering a a fixitReplace be a bit too hard?
Yeah, Doug and I discussed what this would take and agreed that it probably should be a refactoring step and not a fix-it. Things like parameter re-ordering and pulling the code inside the completion handler out into the calling context make it a bit more challenging than a fix-it.

etcwilde commented:
-- forgot to run `git clang-format` before pushing. Nothing semantic changed!

@swift-ci please smoke test

bnbarham commented:
> > Nice one!
> 
> Thanks!
> 
> > Would also offering a a fixitReplace be a bit too hard?
> 
> Yeah, Doug and I discussed what this would take and agreed that it probably should be a refactoring step and not a fix-it. Things like parameter re-ordering and pulling the code inside the completion handler out into the calling context make it a bit more challenging than a fix-it.

It is definitely a more involved fixit than anything we have now, but I was planning on adding it. The way the refactoring works is by replacing the whole call, so that should work fine with a fixit. We'd have to pull out the implementation from refactoring and make it a little more generic, but that shouldn't be *too* bad.

Note that the refactoring doesn't handle parameter re-ordering at the moment, but it wouldn't hurt to make it more robust regardless (it should check it at the very least).

Happy to leave the fixit out of this PR though :)

bnbarham commented on test/attr/attr_completionhandlerasync.swift:
```diff
@@ -137,3 +142,21 @@ func matchingAsyncFunc(value: String) async {} // expected-note{{'matchingAsyncF
 func typecheckFunc9(handler: @escaping () -> Void) {}
+
+// Suggest using async alternative function in async context
+
```
Might be worth adding some tests for the closure case above. Another case is to check we don't output a warning for references of either top level/instance functions.

bnbarham reviewed (commented):
Thanks Evan! I had forgotten to add a radar for this, so I'm glad you remembered :)

etcwilde commented on lib/Sema/TypeCheckConcurrency.h:
```diff
@@ -221,6 +221,13 @@ enum class ConcurrentValueCheck {
 bool checkConcurrentValueConformance(
     ProtocolConformance *conformance, ConcurrentValueCheck check);
 
+/// Check that the async overload
```
No, I just forgot to finish my thought. :) Actually trying to describe what the function does.

etcwilde commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3370,6 +3370,9 @@ ERROR(attr_completion_handler_async_ambiguous_function,none,
       "no corresponding async function named %0", (DeclNameRef))
 
+WARNING(warn_use_async_alternative,none,
+        "consider using asynchronous alternative function",())
```
Are the refactorings something that is Xcode-specific, or can they be used on other platforms/editors? If the refactors only work on the one platform/IDE, it'll look pretty weird for anyone working on other platforms.

bnbarham replied:
Eh, you're right - would look weird in a compiler message from eg. the command line.

etcwilde commented:
Sigh, so it just occurred to me that someone might want to do something like:

```swift
let asyncClosure = { () async -> () in }

@completionHandlerAsync("asyncClosure")
let compHandlerClosure = { (_ handler: @escaping () -> ()) -> () in }

@main struct Main {
  static func main() async {
    compHandlerClosure { print("yuup") }
  }
}
```

Checking this will take a bit more work, so I'll take care of that in a separate PR if that's alright.

bnbarham commented:
> Sigh, so it just occurred to me that someone might want to do something like:
> 
> ```swift
> let asyncClosure = { () async -> () in }
> 
> @completionHandlerAsync("asyncClosure")
> let compHandlerClosure = { (_ handler: @escaping () -> ()) -> () in }
> 
> @main struct Main {
>   static func main() async {
>     compHandlerClosure { print("yuup") }
>   }
> }
> ```
> 
> Checking this will take a bit more work, so I'll take care of that in a separate PR if that's alright.

Does `@completionHandlerAsync` even work in that case at the moment? I thought we restricted it to functions?

etcwilde commented:
> > Sigh, so it just occurred to me that someone might want to do something like:
> > ```swift
> > let asyncClosure = { () async -> () in }
> > 
> > @completionHandlerAsync("asyncClosure")
> > let compHandlerClosure = { (_ handler: @escaping () -> ()) -> () in }
> > 
> > @main struct Main {
> >   static func main() async {
> >     compHandlerClosure { print("yuup") }
> >   }
> > }
> > ```
> > 
> > 
> > Checking this will take a bit more work, so I'll take care of that in a separate PR if that's alright.
> 
> Does `@completionHandlerAsync` even work in that case at the moment? I thought we restricted it to functions?

No, it doesn't right now. I think this starts getting into property-wrapper syntactic territory, so maybe just leave it? I don't know, but it seems like something _someone_ might try to do. I'm going to leave it for now though and pick it back up if it becomes a problem.

bnbarham commented:
> No, it doesn't right now. I think this starts getting into property-wrapper syntactic territory, so maybe just leave it? I don't know, but it seems like something _someone_ might try to do. I'm going to leave it for now though and pick it back up if it becomes a problem.

Sounds good to me!

etcwilde committed:
Requestify @completionHandlerAsync typechecking

This patch puts the type-checking of @completionHandlerAsync behind a
request. We need to be able to request that it gets computed when
checking for warnings of top-level code and global variables that are
closures.

The current order that these get traversed for typechecking is in
declaration ordering, so if we have a global async closure that calls
the completion handler function declared before the completion handler
function declaration, we won't have typechecked the attribute and
resolved the async function declaration yet. This means that we won't
emit the warning, which is bad.

By requestifying it, we can request that the attribute get typechecked
and resolved earlier and verify that the thing typechecks before we do
anything else.

etcwilde committed:
Check global closures

This patch adds checking for usage of completion handler functions with
async alternatives in closures at either the global scope (for
-parse-as-library) or as top-level-decls.

The top-level and global scope are typechecked in the order that they
are declared. This means that if the closure calls a completion
handler function that is declared later, the attribute won't have been
typechecked yet and thus won't have a resolved async function. This
patch updates the warning emitter to typecheck and resolve the attribute
if it needs to. If the thing doesn't typecheck, we don't need emit the
warning at all because the attribute is bad and we tell them to fix it.

DougGregor commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -3060,3 +3060,38 @@ NormalProtocolConformance *GetImplicitConcurrentValueRequest::evaluate(
 }
+
+namespace {
+class UsageChecker : public ASTWalker {
```
I recommend folding this into `MiscDiagnostics.cpp`, which has a whole bunch of these structural checks.

etcwilde committed:
Add iterator for initializers in PatternBindingDecl

This makes it safer and easier to iterate over the initializer
expression in the pattern binding decl. I look forward to refactoring
all the places running an index loop over the `getNumPatternEntries`.

etcwilde committed:
Only traverse closures once

So closures are in a PatternBindingDecl. Since we have an entry-point
into PatternBindingDecl that gets hit for every one of them, we used to
traverse the PBD's that are declared inside of the FunctionDecls twice.

This patch stops us from actually traversing into the PBD from the
checker, and instead pulling out the initializers of interest and
walking each expression separately.

etcwilde committed:
Migrate checker to MiscDiagnostics

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@swift-ci please smoke test

bnbarham reviewed (approved):
Just those couple test additions, but otherwise LGTM

etcwilde committed:
Fix crash with failed typecheck

The walker was crashing if something failed to typecheck since the type
wasn't available. I've added a little test to ensure that not only do we
not crash, but continue checking within the closure.

etcwilde committed:
Adding additional test cases

This patch adds four new test cases

1. Calling the comp-handler function on an instance of this class
2. Calling the comp-handler function on a nested struct instance
3. Calling the comp-handler function on a different struct instance
4. Storing a comp-handler function is just fine

etcwilde commented:
@swift-ci please smoke test

etcwilde commented on test/attr/attr_completionhandlerasync.swift:
```diff
@@ -137,3 +143,61 @@ func matchingAsyncFunc(value: String) async {} // expected-note{{'matchingAsyncF
+      }
+    }
+  }
+}
```
Okay, I think this covers the various test cases for where the functions can be declared and called.

bnbarham replied:
Thanks!
